connectivity/common.c

```diff
@@ -0,0 +1,80 @@
+#include <regex.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "cJSON.h"
+#include "common.h"
+#include "common/logger.h"
+#include "common/ta_errors.h"
+
+#define CONN_LOGGER "connectivity"
+static logger_id_t logger_id;
+
+void conn_logger_init() { logger_id = logger_helper_enable(CONN_LOGGER, LOGGER_DEBUG, true); }
+
+int conn_logger_release() {
+  if (logger_helper_destroy() != RC_OK) {
+    ta_log_error("Destroying logger failed %s.\n", CONN_LOGGER);
+    return EXIT_FAILURE;
+  }
+
+  return 0;
+}
+
+status_t api_path_matcher(char const *const url, char *const regex_rule) {
+  if (regex_rule == NULL) {
+    ta_log_error("%s\n", "SC_HTTP_NULL");
+    return SC_HTTP_NULL;
+  }
+  regex_t reg;
+  regmatch_t pmatch;
+  status_t ret = SC_OK;
+  int reg_flag = REG_EXTENDED;
+
+  if (regcomp(&reg, regex_rule, reg_flag) != 0) {
+    ta_log_error("%s\n", "SC_HTTP_INVALID_REGEX");
+    return SC_HTTP_INVALID_REGEX;
+  }
+  if (regexec(&reg, url, 1, &pmatch, 0) != 0) {
+    // Did not match pattern
+    ret = SC_HTTP_URL_NOT_MATCH;
+  } else {
+    if ((size_t)(pmatch.rm_eo - pmatch.rm_so) != strlen(url)) {
+      ret = SC_HTTP_URL_NOT_MATCH;
+    }
+  }
+
+  regfree(&reg);
+  return ret;
+}
+
+int set_response_content(status_t ret, char **json_result) {
+  int http_ret;
+  if (ret == SC_OK) {
+    return SC_HTTP_OK;
```

Comment:
Should return `MHD_HTTP_OK`

Reply:
`MHD_HTTP_OK` is only used in microhttpd. 
Since MQTT does not use microhttpd, I've changed all original `MHD_XXX` status code to `SC_XXX` status code (they represent same numbers). Is that OK?

Reply:
Then the return type of `set_response_content` should be `status_t`.
But I am wondering why we need to convert one status_t to another? 
The return value of `set_response_content` won't be used and will be overwritten when calling `gossip_channel_set`

Reply:
I'll change the return type to `status_t`
The return values from functions cannot be passed back directly since they are not http status code, so we need to convert them to correspondent codes.

Yes, the return value of `set_response_content` is not used in MQTT, so I'll remove it instead.

Reply:
I found that we did not check if `cJSON_CreateObject` succeeded or not, so the return value of `set_response_content` in MQTT should still be used.

Reply:
Yes, we need the return value of this function.
But the conversion of previous `ret` input to another value does not be used.

Reply:
I still don't find the intention that we change the return code and do nothing.
Can you clarify?

Reply:
I'm not sure what you're talking about, do you mean we didn't use the return value of `set_response_content`? In `connectivity/mqtt/duplex_callback.c:75`, we do use the return code.
```
if (ret == SC_CCLIENT_JSON_CREATE) {
```
However, since we don't have status code in MQTT connection, the return value from `set_response_content` is ONLY used to determine the error, we did not use the returned status code such as `200`, `400`, `404`, etc.

In http connection, on the other hand, we need to set the return value from functions to correspondent status codes, e.g.  `SC_CCLIENT_NOT_FOUND` and `SC_MAM_NOT_FOUND` maps to `404 Not Found`, `SC_CCLIENT_JSON_KEY` and `SC_MAM_NO_PAYLOAD` maps to `400 Bad Request`

Reply:
This point is a good one

Reply:
If the conversion to HTTP status code in MQTT connection is needed, should we utilize the status code? C.c @HowJMay

Reply:
In mqtt we don't really care about the number of status code, but we need a number to identify each error

Reply:
I'll print out the request and correspondent error code after setting response content (either `SC_OK` or other error code).